core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -55,7 +55,7 @@ object TiBatchWrite {
     new TiBatchWrite(df, tiContext, options, regionSplitNumber, enableRegionPreSplit).write()
 }
 
-class TiBatchWrite(@transient val df: DataFrame,
+class TiBatchWrite(@transient var df: DataFrame,
```

Comment:
I don't think it is a good practice because your DataFrame should be a static input. You may use a local dataframe to contain its changes.